fdbserver/DataDistribution.actor.cpp

```diff
@@ -755,68 +755,45 @@ struct DDTeamCollection : ReferenceCounted<DDTeamCollection> {
 
 			int64_t bestLoadBytes = 0;
 			Optional<Reference<IDataDistributionTeam>> bestOption;
-			std::vector<std::pair<int, Reference<IDataDistributionTeam>>> randomTeams;
-			std::set< UID > sources;
-
+			std::vector<Reference<IDataDistributionTeam>> randomTeams;
+			std::set<UID> completeSources;
+			for( int i = 0; i < req.completeSources.size(); i++ ) {
+				completeSources.insert( req.completeSources[i] );
+			}
 			if( !req.wantsNewServers ) {
-				std::vector<Reference<IDataDistributionTeam>> similarTeams;
-				bool foundExact = false;
-
-				for( int i = 0; i < req.sources.size(); i++ )
-					sources.insert( req.sources[i] );
-
-				for( int i = 0; i < req.sources.size(); i++ ) {
-					if( self->server_info.count( req.sources[i] ) ) {
-						auto& teamList = self->server_info[ req.sources[i] ]->teams;
+				int bestSize = 0;
+				for( int i = 0; i < req.completeSources.size(); i++ ) {
+					if( self->server_info.count( req.completeSources[i] ) ) {
+						auto& teamList = self->server_info[ req.completeSources[i] ]->teams;
 						for( int j = 0; j < teamList.size(); j++ ) {
-							if( teamList[j]->isHealthy() && (!req.preferLowerUtilization || teamList[j]->hasHealthyFreeSpace())) {
-								int sharedMembers = 0;
-								for( const UID& id : teamList[j]->getServerIDs() )
-									if( sources.count( id ) )
-										sharedMembers++;
-
-								if( !foundExact && sharedMembers == teamList[j]->size() ) {
-									foundExact = true;
-									bestOption = Optional<Reference<IDataDistributionTeam>>();
-									similarTeams.clear();
-								}
-
-								if( (sharedMembers == teamList[j]->size()) || (!foundExact && req.wantsTrueBest) ) {
-									int64_t loadBytes = SOME_SHARED * teamList[j]->getLoadBytes(true, req.inflightPenalty);
-									if( !bestOption.present() || ( req.preferLowerUtilization && loadBytes < bestLoadBytes ) || ( !req.preferLowerUtilization && loadBytes > bestLoadBytes ) ) {
-										bestLoadBytes = loadBytes;
-										bestOption = teamList[j];
-									}
+							bool found = true;
+							auto serverIDs = teamList[j]->getServerIDs();
+							for( int k = 0; k < teamList[j]->size(); k++ ) {
+								if( !completeSources.count( serverIDs[k] ) ) {
+									found = false;
+									break;
 								}
-								else if( !req.wantsTrueBest && !foundExact )
-									similarTeams.push_back( teamList[j] );
+							}
+							if(found && teamList[j]->isHealthy() && teamList[j]->size() > bestSize) {
+								bestOption = teamList[j];
+								bestSize = teamList[j]->size();
 							}
 						}
+						break;
```

Comment:
Suppose `completeSources` has servers: s1, s2, s3, s4, s5;  none of s1's teams is a subset of {s1, s2, s3, s4, s5}, which means `found = false` and `bestOption` is not found.
However, is it possible that s2 has a team {s2, s3, s4}? In this case, the `bestOption` exists but the for-loop at Line 765 fails to find it.

Reply:
I guess also you could have a situation where s1 has a valid team, but there exists a larger one that doesn't include it.